databricks/koalas/typedef.py

```diff
@@ -235,55 +236,162 @@ def clean_fun(*args2):
-        # Extract the signature arguments from this function.
-        spec = getfullargspec(f)
-        rtype = None
-        return_sig = spec.annotations.get("return", None)
-        if not (return_col or return_sig):
-            raise ValueError(
-                "Missing type information. It should either be provided as an argument to "
-                "pandas_wrap, or as a python typing hint")
-        if return_col is not None:
-            rtype = _to_stype(return_col)
-        if return_sig is not None:
-            rtype = _to_stype(return_sig)
-
-        # Extract the input signatures, if any:
-        sig_args = []
-        sig_kwargs = {}
-        for key in spec.args:
-            t = spec.annotations.get(key, None)
-            if t is not None:
-                dt = _to_stype(t)
-            else:
-                dt = _Unknown(None)
-            sig_kwargs[key] = dt
-            sig_args.append(dt)
-        f.sig_return = rtype
-        f.sig_args = sig_args
-        f.sig_kwargs = sig_kwargs
-        return make_fun(f, *args, **kwargs)
-    return wrapper
+    wrapped_udf = pandas_udf(clean_fun, returnType=return_type)
+    name_tokens = []
+    spark_col_args = []
+    for col in col_args:
+        if col is not None:
+            spark_col_args.append(col._scol)
+            name_tokens.append(col.name)
+    kw_name_tokens = []
+    for (key, col) in col_kwargs:
+        spark_col_args.append(col._scol)
+        kw_name_tokens.append("{}={}".format(key, col.name))
+    col = wrapped_udf(*spark_col_args)
+    series = Series(data=col, index=index_info, anchor=kdf)
+    all_name_tokens = name_tokens + sorted(kw_name_tokens)
+    name = "{}({})".format(f.__name__, ", ".join(all_name_tokens))
+    series = series.astype(return_type).alias(name)
+    return series
+
+
+def _get_metadata(args, kwargs):
+    from databricks.koalas.series import Series
+    all_cols = ([arg for arg in args if isinstance(arg, Series)]
+                + [arg for arg in kwargs.values() if isinstance(arg, Series)])
+    assert all_cols
+    # TODO: check all the anchors
+    s = all_cols[0]
+    return (s._index_info, s._kdf)
+
+
+def pandas_wrap(_function=None, return_col=None, return_scalar=None):
+    """ This annotation makes a function available for Koalas.
+
+    Spark requires more information about the return types than pandas, and sometimes more
+    information is required. This annotations allows you to seamlessly write functions that
+    work for both pandas and koalas.
+
+    Examples
+    --------
+
+    Wrapping a function with python 3's type annotations:
+
+    >>> from databricks.koalas import pandas_wrap, Col
+    >>> pdf = pd.DataFrame({"col1": [1, 2], "col2": [10, 20]}, dtype=np.int64)
+    >>> df = ks.DataFrame(pdf)
+
+    Consider a simple function that operates on pandas series of integers
+
+    >>> def fun(col1):
+    ...     return col1.apply(lambda x: x * 2) # Arbitrary pandas code.
+    >>> fun(pdf.col1)
```

Comment:
@thunterdb, actually, this looks more like Pandas's `apply`. How about we add `apply` with the restriction that it should be wrapped with `pandas_wrap`? I will take a look for that separately and see if that's possible if you think it's a good idea.